File: week1/credit/credit.c
// CS50x Week 1 pset - determines type of credit card if valid

#include <cs50.h>
#include <stdio.h>
#include <math.h>

int digitSum(int n);

int main(void)
{
    // Prompt user for input
    long cardNum = get_long("Number: ");

    // TODO: Validation 1: checksum with Luhn's algorithm
    long temp = cardNum;
    int evenSum = 0;
    int oddSum = 0;
    int i = 0;
    do
    {
        int digit = temp % 10;
        temp /= 10;
        if (i % 2 == 0)
        {
            evenSum += digit;
        }
        else
        {
            oddSum += digitSum(digit * 2);
        }
        i++;
    }
    while (temp > 0);

    // Exit if checksum failed
    if ((evenSum + oddSum) % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Get # of digits
    int digitNo = 0;
    long num = cardNum;
    do
    {
        num /= 10;
        digitNo++;
    }
    while (num > 0);

    // Get first two digits of card number
    int firstTwo = cardNum / pow(10, digitNo - 2);
    
    // Validation 2: check for validity and card type
    if (digitNo == 15 && (firstTwo == 34 || firstTwo == 37))
    {
        printf("AMEX\n");
    }
    else if (digitNo == 16 && (firstTwo > 50 && firstTwo < 56))
    {
        printf("MASTERCARD\n");
    }
    else if ((digitNo == 13 || digitNo == 16) && (firstTwo > 39 && firstTwo < 50))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// Sum together every digit in an int
int digitSum(int n)
{
    int temp = n;
    int sum = 0;
    do
    {
        int d = temp % 10;
        temp /= 10;
        sum += d;
    }
    while (temp > 0);

    return sum;
}

File: week6/pset6/sentimental-credit/credit.py
# Credit!
# https://docs.python.org/3/howto/regex.html#regex-howto

from sys import exit
import re
from math import floor

while True:
    try:
        n = int(input("Number: "))
        break
    except ValueError:
        continue

string = str(n)
counter = 0
even_sum = 0
odd_sum = 0
passes_Luhn = False

while n > 0:
    digit = n % 10
    n = floor(n / 10)
    if counter % 2 == 0:
        even_sum += digit
    else:
        temp = digit * 2
        while temp > 0:
            odd_sum += temp % 10
            temp = floor(temp / 10)
    counter += 1

if (even_sum + odd_sum) % 10 == 0:
    passes_Luhn = True
else:
    print("INVALID")
    exit(1)

AMEX = "3[47]............."
MASTERCARD = "5[1-5].............."
VISA1 = "4............"
VISA2 = "4..............."

AMEX_pattern = re.compile(AMEX)
MASTERCARD_pattern = re.compile(MASTERCARD)
VISA1_pattern = re.compile(VISA1)
VISA2_pattern = re.compile(VISA2)

if AMEX_pattern.search(string) is not None:
    print("AMEX")
elif MASTERCARD_pattern.search(string) is not None:
    print("MASTERCARD")
elif (
    VISA1_pattern.search(string) is not None or VISA2_pattern.search(string) is not None
):
    print("VISA")
else:
    print("INVALID")
